sys/net/gnrc/routing/rpl/srh/gnrc_rpl_srh.c

```diff
@@ -12,13 +12,86 @@
  * @file
  */
 
+#include <string.h>
+#include "net/gnrc/ipv6/netif.h"
 #include "net/gnrc/rpl/srh.h"
 
-ipv6_addr_t *gnrc_rpl_srh_next_hop(gnrc_rpl_srh_t *rh)
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#if ENABLE_DEBUG
+static char addr_str[IPV6_ADDR_MAX_STR_LEN];
+#endif
+
+#define GNRC_RPL_SRH_PADDING(X)     ((X & 0xF0) >> 4)
+#define GNRC_RPL_SRH_COMPRE(X)      (X & 0x0F)
+#define GNRC_RPL_SRH_COMPRI(X)      ((X & 0xF0) >> 4)
+
+int gnrc_rpl_srh_process(ipv6_hdr_t *ipv6, gnrc_rpl_srh_t *rh)
 {
-    /* TODO */
+    if (rh->seg_left == 0) {
+        return EXT_RH_CODE_OK;
+    }
+
+    uint8_t n = (((rh->len * 8) - GNRC_RPL_SRH_PADDING(rh->pad_resv) -
+                 (16 - GNRC_RPL_SRH_COMPRE(rh->compr))) /
+                 (16 - GNRC_RPL_SRH_COMPRI(rh->compr))) + 1;
+    ipv6_addr_t addr = ipv6->dst;
+    ipv6_addr_t tmp = ipv6->dst;
+    uint8_t i, pref_elided, addr_len;
+    uint8_t *addr_vec = (uint8_t *) (rh + 1);
+
+    DEBUG("RPL SRH: %" PRIu8 " addresses in the routing header\n", n);
+
+    if (rh->seg_left > n) {
+        DEBUG("RPL SRH: number of segments left > number of addresses - discard\n");
+        /* TODO ICMP Parameter Problem - Code 0 */
+        return EXT_RH_CODE_ERROR;
+    }
+
+    rh->seg_left--;
+    i = n - rh->seg_left;
+    pref_elided = rh->seg_left ? GNRC_RPL_SRH_COMPRI(rh->compr) : GNRC_RPL_SRH_COMPRE(rh->compr);
+    addr_len = sizeof(ipv6_addr_t) - pref_elided;
+    memcpy(&addr.u8[pref_elided], &addr_vec[(i - 1) * addr_len], addr_len);
```

Comment:
That's right. It is preferable to check `ext->len` comprehensively at `gnrc_ipv6_ext_demux` or somewhere.